PGM/src/main/java/tc/oc/pgm/shop/SimplePurchaseTracker.java

```diff
@@ -0,0 +1,118 @@
+import java.util.Set;
+
+/**
+ * Simple implementation of the {@link PurchaseTracker}.
+ */
+class SimplePurchaseTracker implements PurchaseTracker {
+    private final Map<MatchPlayer, PaymentStrategy> individualStrategies;
+    private final Map<Competitor, PaymentStrategy> competitorStrategies;
+    private final Set<PaymentStrategy> globalStrategies;
+
+    private final boolean crossParty;
+    private final boolean persistent;
+
+    public SimplePurchaseTracker(boolean crossParty, boolean persistent) {
+        this.individualStrategies = Maps.newHashMap();
+        this.competitorStrategies = Maps.newHashMap();
+        this.globalStrategies = Sets.newHashSet();
+        this.crossParty = crossParty;
+        this.persistent = persistent;
+    }
+
+    @Override
+    public Filter defaultFilter(Purchasable purchasable) {
+        return purchasable.getPurchaseFilter();
+    }
+
+    @Override
+    public void startIndividualStrategy(MatchPlayer player, Purchasable purchasable, Filter filter) {
+        PlayerPaymentStrategy strategy = new PlayerPaymentStrategy(purchasable, filter, player, crossParty, persistent);
+        individualStrategies.put(player, strategy);
+        contribute(strategy, player);
+    }
+
+    @Override
+    public void startCompetitorStrategy(Competitor competitor, MatchPlayer player, Purchasable purchasable, Filter filter) {
+        CompetitorPaymentStrategy strategy = new CompetitorPaymentStrategy(purchasable, filter, competitor);
+        competitorStrategies.put(competitor, strategy);
+        contribute(strategy, player);
+    }
+
+    @Override
+    public void startGlobalStrategy(Purchasable purchasable, MatchPlayer player, Filter filter) {
+        GlobalPaymentStrategy strategy = new GlobalPaymentStrategy(purchasable, filter);
+        globalStrategies.add(strategy);
+        contribute(strategy, player);
+    }
+
+    @Override
+    public Optional<PaymentStrategy> getOngoingStrategy(MatchPlayer player, Purchasable purchasable) {
+        return individualStrategies.values().stream().filter(e ->
+            e.getOwner().equals(player) && e.getPurchasable().equals(purchasable)
+        ).findFirst();
+    }
+
+    @Override
+    public Optional<PaymentStrategy> getOngoingStrategy(Competitor competitor, Purchasable purchasable) {
+        return competitorStrategies.values().stream().filter(e ->
+                e.getOwner().equals(competitor) && e.getPurchasable().equals(purchasable)
+        ).findFirst();
+    }
+
+    @Override
+    public Optional<PaymentStrategy> getOngoingStrategy(Purchasable purchasable) {
+        return competitorStrategies.values().stream().filter(e ->
+                e.getPurchasable().equals(purchasable)
+        ).findFirst();
+    }
+
+    @Override
+    public void contribute(PaymentStrategy strategy, MatchPlayer player) {
+        strategy.contribute(player);
+        if (strategy.isComplete()) {
+            switch (strategy.getPurchasable().getType()) {
```

Comment:
i feel like this switch could be part of strategy.contribute method, and let every implementation deal with it how it wants to, instead of a switch